src/lib/FreeC/Backend/Agda/Converter/Type.hs

```diff
@@ -0,0 +1,87 @@
+
+import           Control.Monad                  ( mapM )
+
+import qualified FreeC.Backend.Agda.Syntax     as Agda
+import           FreeC.Backend.Agda.Converter.Free
+                                                ( free
+                                                , applyFreeArgs
+                                                )
+import           FreeC.Environment.LookupOrFail ( lookupAgdaIdentOrFail )
+import           FreeC.Environment.Renamer      ( renameAndDefineAgdaTypeVar )
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.Monad.Converter          ( Converter )
+
+-- | Lifts a single type.
+convertType :: IR.Type -> Converter Agda.Expr
+convertType = dagger
+
+-- | Functions not defined in terms lambdas (@f x = …@ not @f = \x -> …@) aren't
+--   evaluated until fully applied, i.e. cannot be bottom. Therefore interleaving
+--   monadic layers isn't needed.
+--
+--   > (τ₁ → τ₂ → … → τₙ)' = τ₁' → Free τ₂' → … → Free τₙ'
+convertFunctionType :: [IR.Type] -> Converter Agda.Expr
+convertFunctionType types = foldr1 Agda.fun <$> mapM dagger types
+
+-- | Constructors aren't evaluated until fully applied. Furthermore the return
+--   type of a constructor for a data type D has to be D and therefore cannot be
+--   lifted.
+--
+--   > (τ₁ → τ₂ → … → τₙ → D)' = τ₁' → Free τ₂' → … → τₙ' → D
+convertConstructorType :: [IR.Type] -> Converter Agda.Expr
+convertConstructorType types =
+  -- We can use the @star@ translation for the data type, because only the name
+  -- and the application of type and @Size@ variables have to be translated.
+  Agda.fun <$> convertFunctionType (init types) <*> star (last types)
+
+-- | Utility function for introducing a new Agda type variable to the current
+--   scope.
+renameAgdaTypeVar :: IR.TypeVarDecl -> Converter Agda.Name
+renameAgdaTypeVar (IR.TypeVarDecl srcSpan name) =
+  Agda.unqualify <$> renameAndDefineAgdaTypeVar srcSpan name
+
+-------------------------------------------------------------------------------
+-- Translations                                                              --
+-------------------------------------------------------------------------------
+
+-- | The dagger translation as describer by Abel et al.
+--   Normally the dagger translations separates mono and poly types, but the
+--   polytype cases are not applicable for our IR.
+--   Note: If polytypes are added in the future (e.g. for type classes) add
+--   their type translation here.
+--
+--   > -- not applicable (polytypes)
+--   > (∀α:κ.σ)’ = (α : κ’) → σ’ -- RankNTypes translate to pi types
+--   > (σ ↦ τ)’ = σ’ → τ’        -- functions handling polytypes e.g.
+--   >                           -- result from type class dictionary translation
+--   > -- all other cases (monotypes):
+--   > τ’ = m τ
+dagger :: IR.Type -> Converter Agda.Expr
+dagger = fmap free . star
+
+-- | The star translation for monotypes as described by Abel et al.
+--
+-- > (σ τ)* = σ* τ*
+-- > (σ → τ)* = σ’ → τ’
```

Comment:
`σ` is often used for substitutions in the documentation. If multiple meta variables of the same type occur, they usually have indices. Thus, I'd write `τ₁` and `τ₂`.